libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -607,17 +699,24 @@ export class QnAMakerDialog extends WaterfallDialog implements QnAMakerDialogCon
             qnaResponse.answers = qna.getLowScoreVariation(qnaResponse.answers);
 
             if (isActiveLearningEnabled && qnaResponse.answers?.length > 1) {
-                const suggestedQuestions: string[] = [];
-
-                qnaResponse.answers.forEach((answer) => {
-                    suggestedQuestions.push(answer.questions[0]);
-                });
-
-                const message = QnACardBuilder.getSuggestionsCard(
-                    suggestedQuestions,
-                    dialogOptions.qnaDialogResponseOptions.activeLearningCardTitle,
-                    dialogOptions.qnaDialogResponseOptions.cardNoMatchText
-                );
+                const suggestedQuestions = qnaResponse.answers.map((answer) => answer.questions[0]);
+
+                const message = this.suggestionsActivityFactory
+                    ? this.suggestionsActivityFactory(
+                          suggestedQuestions,
+                          dialogOptions.qnaDialogResponseOptions.cardNoMatchText
+                      )
+                    : QnACardBuilder.getSuggestionsCard(
+                          suggestedQuestions,
+                          dialogOptions.qnaDialogResponseOptions.activeLearningCardTitle,
+                          dialogOptions.qnaDialogResponseOptions.cardNoMatchText
+                      );
+
+                // QnACardBuilder.getSuggestionsCard is a public static method and can be overridden by the developer.
+                // If either of these methods do not return an Activity, throw a TypeError.
+                if (!isSuggestionsActivity(message)) {
```

Comment:
I left a note above, but using a test and then throwing is an antipattern - that's what type assertions are for.

Reply:
Good feedback, I had to use `asserts.object` because what we have in botframework-schema is too strict at this level in the application.

isActivity returns false for the result of `MessageFactory.suggestedActions(['suggestion1', 'suggestion2'], 'noMatchCardText')` which is a Partial<Activity> type that TurnContext.sendActivity() accepts.

Information required to send the Activity to the right conversation (e.g. serviceUrl, conversation references, recipient) are populated by the SDK in TurnContext.sendActivity when said info is missing.